[util] Re-clone Vendor without changing Revision

lenary opened the pull request:
This commit adds the `--no-update` command line flag to `util/vendor.py`
which reads the version of the repository in the `.lock.hjson` file and
uses that version when cloning the repository rather than the version
in `.vendor.hjson`. This should mean the cloned version of the
repository is exactly the same version as what was cloned the last time
the repository was updated.

This allows developers to update patches and file exclusion rules in the
`.vendor.hjson` file without having to integrate any upstream changes.

---

This is the second time I've needed this functionality, so I went ahead and
added it.

imphil commented:
Thanks @lenary, that's much needed functionality.

I'd argue that going the other way around is more natural and in line with other tools like bundler, composer, pip, etc.: make no updates by default, and only update if an `--update` flag is given.

Additionally, this needs changes to the documentation (probably both reference docs and user guide).

lenary commented:
I have rewritten this to use `--update` if updates are wanted. I think I have this correct, and `--refresh-patches` is still doable even if `--update` is not passed to vendor.

I have also updated the docs.

I'm working on running this in CI, now that vendoring is reproducible. I'm not sure how viable this is, but that's what #2045 does (a WIP at the moment).

lenary commented:
This got a lot of pushes because running it in CI (#2045) kept finding python `pathlib` issues. I think they're all solved now, or at least solved enough for the moment.

imphil commented on util/vendor.py:
```diff
@@ -340,46 +358,82 @@ def main(argv):
+                )
+                log.fatal("Locked Version: %s", rev_to_clone)
+                log.fatal("Re-cloned Version: %s", upstream_new_rev)
+                raise SystemExit(1)
```
Can you put this check into clone_git_repo?, optionally as an assert? It's a violation of the postcondition of `clone_git_repo`.

lenary replied:
This cannot go into `clone_git_repo`: if you're using `--update`, the `rev` argument to `clone_git_repo` will likely be something like `master`, but the return value is always a sha. I'll try to work out how to make this check slightly less invasive.

imphil commented on util/vendor.py:
```diff
@@ -340,46 +358,82 @@ def main(argv):
+            log.warning(
+                "Unable to read lock file %s. Assuming this is the first import.",
+                str(lock_file_path))
+            first_import = True
```
You can simplify the logic with "`--update` is implied if no lock file exists".

```
ignore_lock_file = args.update or log_file_doesnt_exist
```

and then continue as usual

lenary replied:
This is a little more complex, because `first_import` was used by `get_shortlog`. I've tried to match the suggested logic, along with the recommendation about `import_desc` below, but one of the compilations is that the `lock` information is needed when finding out what changed.

The end result is the logic will need re-reviewing, because it is simpler (update is now assumed if the lock file does not exist), but it has also changed quite a bit.

imphil commented on util/vendor.py:
```diff
@@ -388,31 +442,27 @@ def main(argv):
-            log.warning(
-                "No lock file exists. Unable to get the log of changes.")
-        elif lock['upstream']['url'] != desc['upstream']['url']:
+            log.warning("Initial import: unable to summarise changes.")
```
```suggestion
            log.warning("Initial import: unable to summarize changes.")
```

:smile:

imphil commented on util/vendor.py:
```diff
@@ -340,46 +358,82 @@ def main(argv):
+            rev_to_clone = desc['upstream']['rev']
+        else:
+            url_to_clone = lock['upstream']['url']
+            rev_to_clone = lock['upstream']['rev']
```
only_subdir is also part of the lock file and needs to be used from there.

There is room for discussion if things like the excludes and the applied patches should be part of the lock file as well and applied from there, but that's probably beyond this PR.

To prepare for that, I'd for something like

```
import_desc = desc.copy()
import_desc['upstream'] = lock['upstream']
```

and use import_desc in the code below.

lenary replied:
I wasn't sure about `only_subdir`, but I think this change does make the code clearer, so I've gone with it. It helped make the documentation clearer too.

lenary commented on util/vendor.py:
```diff
@@ -224,13 +224,25 @@ def apply_patch(basedir, patchfile, strip_level=1):
-        'git', 'clone', '--no-single-branch', '-b', rev, repo_url, clone_dir
-    ]
+    # We might be passed a rev or a sha here
+    is_sha = len(rev) == 40
```
Sure, sounds good.

lenary commented on util/vendor.py:
```diff
@@ -349,19 +368,45 @@ def main(argv):
+        if args.update:
+            log.fatal("Cannot update without existing lock file %s.", lock_file_path)
+            log.fatal("Retry without the '--update' flag if this is an initial import.")
+            raise SystemExit(1)
```
Ok, that makes sense.

lenary commented on util/vendor.py:
```diff
@@ -349,19 +368,45 @@ def main(argv):
+            if upstream_new_rev == lock['upstream']['rev']:
+                log.fatal("Upstream Revision has not changed.")
+                log.fatal("This could be because the rev in %s was not updated", desc_file_path)
+                raise SystemExit(1)
```
Ok, Error has gone.

lenary commented on doc/rm/vendor_in_tool.md:
```diff
@@ -100,10 +101,32 @@ In the example vendor description file below, the mpsse directory is populated f
 
+## Updating and The Vendor Lock File
+
+In order to document which version of a repositoy has been cloned and committed to the repository with the vendor tool, a vendor lock file is stored in `vendor/<vendor>_<name>.lock.hjson`. This contains only the upstream information, including the URL and the exact git revision that was cloned.
```
Oops, intended to do that, missed this sentence. Fixed now.

imphil commented on util/vendor.py:
```diff
@@ -340,79 +352,94 @@ def main(argv):
-        '.lock.hjson')
-    try:
-        with open(lock_file_path, 'r') as f:
+    if lock_file_path.is_file():
```
The try/catch approach is more common in Python and race-condition/corner-case-safer ("it's easier to ask for forgiveness than permission"), but fine either way.

lenary replied:
Addressed.

imphil commented on util/vendor.py:
```diff
@@ -340,79 +352,94 @@ def main(argv):
+        if not args.update:
+            log.warning("Updating upstream repo as lock file %s not found.",
+                        str(lock_file_path))
+            args.update = True
```
nit: I wouldn't modify `args.update`, but use a new variable for it.

lenary replied:
I think using `args.update` is easier at the moment. We can always revisit this.

imphil commented on util/vendor.py:
```diff
@@ -340,79 +352,94 @@ def main(argv):
+        if not args.update:
+            if upstream_new_rev != lock['upstream']['rev']:
+                log.fatal(
+                    "Revision Mismatch. Unable to re-clone locked version of repository."
```
```suggestion
                    "Revision mismatch. Unable to re-clone locked version of repository."
```

lenary replied:
Addressed and also slightly reworded and recapitalised the fatal errors below.
